far/macroapi.cpp

```diff
@@ -1204,31 +1251,62 @@ void KeyMacro::CallFar(intptr_t CheckCode, FarMacroCall* Data)
 		case MCODE_F_MENU_GETEXTENDEDDATA: // T=Menu.GetItemExtendedData([hDlg,][Pos])
 		{
 			const auto Params{ api.parseParams(2) };
-			auto Nidx = 0;
-			Dialog* Dlg{};
-			if(Params[0].isDialog())
+			auto ParamIter{ Params.cbegin() };
+			auto Dlg{ get_optional_dialog_param(ParamIter, GetArea(), CurrentWindow) };
+			const auto N{ get_optional_integer_param(ParamIter).value_or(0) - 1 };
+
+			if (!Dlg)
 			{
-				Nidx = 1;
-				Dlg = Params[0].asDialog();
+				api.PushNil();
+				return;
 			}
-			else if(IsMenuOrDialogArea(GetArea()) && CurrentWindow)
+
+			if (VMenu::extended_item_data ExtendedData; Dlg->VMProcess(CheckCode, &ExtendedData, N) == 1)
 			{
-				Dlg = dynamic_cast<Dialog*>(CurrentWindow.get());
+				api.PushTable();
+				for (const auto& [Key, Value] : ExtendedData)
+				{
+					api.SetField(Key, Value, -3);
```

Comment:
What is -3?
Maybe introduce some named constant or add a comment?

Reply:
> What is -3?
Maybe introduce some named constant or add a comment?

It is table position in Lua stack.
Stack position is often expressed as a negative value relative to stack top.
When we want to set a key-value pair in a table we first push the key then push the value.
If initially the table was on Lua stack top its stack position was -1.
After key and value were pushed the table position became -3.

Reply:
Thank you for the explanation, @shmuz.

@alabuzhev, as you can see, it is not a magic constant. I'll add a comment.